Use Spark shared Hadoop configuration

andrei-ionescu opened the pull request:
### What is the context for this pull request?

 - **Tracking Issue**: #322 
 - **Parent Issue**: N/A
 - **Dependencies**: N/A
 - Fixes: #322 

### What changes were proposed in this pull request?

Use `spark.sessionState.newHadoopConf()` instead of `new Configuration()` in the sections where there is interaction with the file system.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

1. All already present tests are passing
2. Locally & Databricks Runtime 5.5 LTS & 6.4 tests with ADL storage

imback82 commented on src/main/scala/com/microsoft/hyperspace/util/FileUtils.scala:
```diff
@@ -68,8 +69,9 @@ object FileUtils {
    */
-  def createDirectory(dirPath: Path): Unit = {
-    val fileSystem = dirPath.getFileSystem(new Configuration)
+  def createDirectory(dirPath: Path,
+      hadoopConfiguration: Configuration = new Configuration): Unit = {
+    val fileSystem = dirPath.getFileSystem(hadoopConfiguration)
```
Are you using `scalafmt`? I get the following if I auto-format:
```scala
  def createDirectory(
      dirPath: Path,
      hadoopConfiguration: Configuration = new Configuration): Unit = {
    val fileSystem = dirPath.getFileSystem(hadoopConfiguration)
    if (!fileSystem.exists(dirPath)) {
      fileSystem.mkdirs(dirPath)
    }
  }
```

imback82 replied:
Can you check other places as well?

andrei-ionescu replied:
@imback82 I'm using the `scalastyle-config.xml` file present in the project. 

I tried setting **scalafmt** but it needs a `scalafmt.conf` config file. Is that present somewhere in the project?

sezruby replied:
@andrei-ionescu Please use this: https://github.com/microsoft/hyperspace/blob/master/dev/.scalafmt.conf

andrei-ionescu replied:
@imback82 Ultimately I did found the `.scalafmt.conf` file in `dev` folder. Will try to apply it.

andrei-ionescu replied:
There seems to be some sort of misconfiguration in the scalafmt conf file because when I run it with optimise imports on a file it reorders the imports putting the `com.microsoft.hyperspace.*` at the top.

imback82 commented on src/main/scala/com/microsoft/hyperspace/index/factories.scala:
```diff
@@ -20,31 +20,32 @@ import org.apache.hadoop.conf.Configuration
-  private val configuration = new Configuration
-
-  override def create(path: Path): FileSystem = path.getFileSystem(configuration)
+  override def create(path: Path, configuration: Configuration = new Configuration): FileSystem =
```
Looks like these `*Impl` classes will always be created with default configurations right (since I don't see changes when `CachingIndexCollectionManager` or `IndexCollectionManager` is created). Does it make sense to use the same `Configuration` setting everywhere?

andrei-ionescu replied:
👍 You're right! I missed the places where the `create` methods of `IndexLogManagerFactory` & `IndexDataManagerFactory`.

imback82 commented on src/main/scala/com/microsoft/hyperspace/index/factories.scala:
```diff
@@ -20,31 +20,32 @@ import org.apache.hadoop.conf.Configuration
 
 trait FileSystemFactory {
-  def create(path: Path): FileSystem
+  def create(path: Path, configuration: Configuration = new Configuration): FileSystem
```
Can we remove default parameter value from `trait`? (I am not a big fan of it since derived class can override the value and can cause some confusion).

andrei-ionescu replied:
👍

andrei-ionescu commented:
@imback82, @sezruby Can you have another look on this? Thanks.

sezruby reviewed (commented):
LGTM Thanks!

andrei-ionescu commented:
Thanks @sezruby! @imback82 can you have a look also?

imback82 commented on src/main/scala/com/microsoft/hyperspace/index/IndexCollectionManager.scala:
```diff
@@ -34,11 +34,13 @@ class IndexCollectionManager(
   override def create(df: DataFrame, indexConfig: IndexConfig): Unit = {
-    val indexPath = PathResolver(spark.sessionState.conf).getIndexPath(indexConfig.indexName)
-    val dataManager = indexDataManagerFactory.create(indexPath)
+    val indexPath = PathResolver(spark.sessionState.conf,
+      spark.sessionState.newHadoopConf()).getIndexPath(indexConfig.indexName)
+    val dataManager = indexDataManagerFactory.create(indexPath,
+      spark.sessionState.newHadoopConf())
```
Can you run scalafmt on this file? I see a different formatting:
```scala
    val indexPath = PathResolver(spark.sessionState.conf, spark.sessionState.newHadoopConf())
      .getIndexPath(indexConfig.indexName)
    val dataManager =
      indexDataManagerFactory.create(indexPath, spark.sessionState.newHadoopConf())
```

andrei-ionescu replied:
👍

imback82 commented on src/main/scala/com/microsoft/hyperspace/index/IndexCollectionManager.scala:
```diff
@@ -58,16 +60,20 @@ class IndexCollectionManager(
     withLogManager(indexName) { logManager =>
-      val indexPath = PathResolver(spark.sessionState.conf).getIndexPath(indexName)
-      val dataManager = indexDataManagerFactory.create(indexPath)
+      val indexPath = PathResolver(spark.sessionState.conf,
+        spark.sessionState.newHadoopConf()).getIndexPath(indexName)
+      val dataManager = indexDataManagerFactory.create(indexPath,
+        spark.sessionState.newHadoopConf())
```
ditto

andrei-ionescu replied:
👍

imback82 commented on src/main/scala/com/microsoft/hyperspace/actions/OptimizeAction.scala:
```diff
@@ -135,11 +135,12 @@ class OptimizeAction(
     // the list of newly created index files.
-    val absolutePath = PathUtils.makeAbsolute(indexDataPath)
-    val newContent = Content.fromDirectory(absolutePath, fileIdTracker)
+    val absolutePath = PathUtils.makeAbsolute(indexDataPath.toString)
+    val newContent = Content.fromDirectory(absolutePath, fileIdTracker,
+      hadoopConfiguration = spark.sessionState.newHadoopConf())
```
ditto on the formatting

andrei-ionescu replied:
👍

imback82 reviewed (commented):
@andrei-ionescu could you run scalafmt on the files you modified? If it doesn't work, please let me know and I can push the changes to your branch. Thanks!

imback82 commented on src/main/scala/com/microsoft/hyperspace/actions/OptimizeAction.scala:
```diff
@@ -135,11 +135,12 @@ class OptimizeAction(
     // the list of newly created index files.
-    val absolutePath = PathUtils.makeAbsolute(indexDataPath)
-    val newContent = Content.fromDirectory(absolutePath, fileIdTracker)
+    val absolutePath = PathUtils.makeAbsolute(indexDataPath.toString)
+    val newContent = Content.fromDirectory(absolutePath, fileIdTracker,
+      hadoopConfiguration = spark.sessionState.newHadoopConf())
```
Shall we do the following? It's a bit confusing when to use hadoop config or not:
```scala
val hadoopConf = spark.sessionState.newHadoopConf())
val absolutePath = PathUtils.makeAbsolute(indexDataPath, hadoopConf)
val newContent = Content.fromDirectory(absolutePath, fileIdTracker, hadoopConfiguration = hadoopConf)
```

andrei-ionescu replied:
👍

imback82 commented on src/main/scala/com/microsoft/hyperspace/util/PathUtils.scala:
```diff
@@ -20,10 +20,11 @@ import org.apache.hadoop.conf.Configuration
 
-  def makeAbsolute(path: Path): Path = {
-    val fs = path.getFileSystem(new Configuration)
+  def makeAbsolute(path: Path, hadoopConfiguration: Configuration): Path = {
```
Any reason this doesn't have a default parameter whereas the above one does?

andrei-ionescu replied:
Yes - **overloading**. The `def makeAbsolute(path: String, hadoopConfiguration: Configuration = new Configuration)` is overloading the one at line 26. Compiler complains when both have default parameters:

```
PathUtils.scala:22: in object PathUtils, multiple overloaded alternatives of 
makeAbsolute define default arguments
```

andrei-ionescu commented:
@imback82 I did run `scalafmt` on the code I changed, I integrated your feedback and rebased it on top of master. 

BTW, running `scalafmt` on the whole file moves the private methods to different locations and shuffles the imports. That why I did apply it only on the code I changed.

imback82 commented on src/main/scala/com/microsoft/hyperspace/actions/CreateActionBase.scala:
```diff
@@ -52,7 +52,7 @@ private[actions] abstract class CreateActionBase(dataManager: IndexDataManager)
       indexConfig: IndexConfig,
       path: Path): IndexLogEntry = {
-    val absolutePath = PathUtils.makeAbsolute(path)
+    val absolutePath = PathUtils.makeAbsolute(path.toString, spark.sessionState.newHadoopConf())
```
To minimize changes:
```suggestion
    val absolutePath = PathUtils.makeAbsolute(path, spark.sessionState.newHadoopConf())
```

andrei-ionescu replied:
👍

imback82 commented on src/main/scala/com/microsoft/hyperspace/actions/OptimizeAction.scala:
```diff
@@ -135,11 +135,13 @@ class OptimizeAction(
-    val absolutePath = PathUtils.makeAbsolute(indexDataPath)
-    val newContent = Content.fromDirectory(absolutePath, fileIdTracker)
+    val hadoopConf = spark.sessionState.newHadoopConf()
+    val absolutePath = PathUtils.makeAbsolute(indexDataPath.toString)
```
```suggestion
    val absolutePath = PathUtils.makeAbsolute(indexDataPath, hadoopConf)
```
as suggested here: https://github.com/microsoft/hyperspace/pull/327#discussion_r564142248

andrei-ionescu replied:
👍 Good catch. I missed this one.

imback82 commented on src/main/scala/com/microsoft/hyperspace/index/IndexCollectionManager.scala:
```diff
@@ -34,11 +34,13 @@ class IndexCollectionManager(
   override def create(df: DataFrame, indexConfig: IndexConfig): Unit = {
-    val indexPath = PathResolver(spark.sessionState.conf).getIndexPath(indexConfig.indexName)
-    val dataManager = indexDataManagerFactory.create(indexPath)
+    val indexPath = PathResolver(spark.sessionState.conf, spark.sessionState.newHadoopConf())
```
shall we create a variable for `spark.sessionState.newHadoopConf()` and reuse it? I see `newHadoopConf` three times here.

imback82 replied:
Looks like it's used many times in this class.
We can either create one `lazy val` at the class level or create a variable at a function level if caching at class level is a concern for staleness.

andrei-ionescu replied:
Yes, caching may be an issue as the shared spark's Hadoop configuration may change and we want to have it updated in Hyperspace too. I'll create method variables.

imback82 commented on src/main/scala/com/microsoft/hyperspace/index/IndexLogManager.scala:
```diff
@@ -160,4 +155,12 @@ class IndexLogManagerImpl(indexPath: Path) extends IndexLogManager with Logging
     }
   }
+
+  private def getLog(path: Path): Option[LogEntry] = {
```
nit: can we move this up to minimize changes?

andrei-ionescu replied:
Yes. Sure. I'll put it back. This un-wanted change is due to applying `scalafmt` on the whole file.

imback82 reviewed (commented):
LGTM (few minor comments), thanks @andrei-ionescu!

imback82 commented:
> BTW, running `scalafmt` on the whole file moves the private methods to different locations and shuffles the imports. That why I did apply it only on the code I changed.

Weird. Are you using the this file? https://github.com/microsoft/hyperspace/blob/master/dev/.scalafmt.conf

What I do is to import the above file in IntellijJ and `Code` -> `Reformat code`. I will write up something on how to fix the imports rearrange in IntelliJ.

andrei-ionescu commented:
@imback82 Regarding the scala style and scalafmt. I did import the `/dev/.scalafmt.conf` file in IntelliJ's settings. What I do is this: on the file that change I do `Reformat file...` (`Cmd + Alt + Shift + L`) and from the dialog that opens I choose: `Whole file` + `Rearrange the code` + `Code cleanup`. I think [scala-coding-style.md guide](https://github.com/microsoft/hyperspace/blob/master/docs/coding-guidelines/scala-coding-style.md) needs some more information.

andrei-ionescu commented:
@imback82 I integrated your feedback. Thank you for it!
